deptry/virtualenv_finder.py

```diff
@@ -0,0 +1,100 @@
+) -> bool:
+    """Determine if executed by the interpreter in the project's virtual environment
+
+    If we are executed from virtual environment, the `prefix` argument
+    will be set to the virtual environment's directory, whereas
+    `base_prefix` will point to the global Python installation
+    used to create the virtual environment.
+
+    The `active_virtual_env` arguments holds the value of the
+    VIRTUAL_ENV environment variable, that tells with good reliability
+    that a virtual environment has been activated in the current shell.
+    """
+
+    # Gobal installation
+    if prefix == base_prefix:
+        return False
+
+    # No virtualenv has been activated. Unless the project name is in
+    # the interpreter path, assume we are not in the project's virtualenv
+    if not active_virtual_env:
+        return project_name in prefix
+
+    # A virtualenv has been activated. But if `deptry` was installed gloabally using
+    # `pipx`, we could be running in another installation.
+    return active_virtual_env == prefix
+
+
+def guess_virtualenv_site_packages(project_root: Path, active_virtual_env: Optional[str] = None) -> Optional[Path]:
+    """Try to locate a project's virtualenv packages using environment and popular locations"""
+    site_packages = None
+    possible_roots = [
+        Path("~/.virtualenvs").expanduser() / project_root.name,
+        project_root / ".venv",
+    ]
+    if active_virtual_env:
+        possible_roots.append(Path(active_virtual_env))
+
+    while not site_packages and possible_roots:
+        site_packages = find_site_packages_in(possible_roots.pop())
+
+    return site_packages
+
+
+def install_distribution_finder(site_packages: Path) -> None:
+    path = [str(site_packages.absolute()), *sys.path]
+
+    class VirtualenvDistributionFinder(metadata.MetadataPathFinder):
+        @classmethod
+        def find_distributions(cls, context):
+            context = metadata.DistributionFinder.Context(name=context.name, path=path)
+            return super().find_distributions(context)
+
+    sys.meta_path.insert(0, VirtualenvDistributionFinder())
+
+
+def maybe_install_finder(project_root: Path) -> None:
+    """If need be, add poject virtualenv site packages to metadata search path"""
+
+    project_name = project_root.absolute().name
+    virtual_env = os.environ.get("VIRTUAL_ENV")
+
+    # If we are likely executed by the python interpreter of the
+    # project's virtual envrionment, nothing to do!
+    if running_in_project_virtualenv(
+        project_name, prefix=sys.prefix, base_prefix=sys.base_prefix, active_virtual_env=virtual_env
+    ):
+        return
+
+    # Try to locate the project's virtual environment site-packages
+    # and add it to the dependency metadata search path.
+    site_packages = guess_virtualenv_site_packages(project_root, virtual_env)
+    if site_packages:
+        logging.warning("Assuming virtual environment for project %s is %s", project_name, site_packages)
```

Comment:
Can you change this to an [f-string](https://peps.python.org/pep-0498/)? Maybe a bit nitpicky, but that's what is used in the rest of this project.

Reply:
Just a note that [it's considered bad practice](https://docs.python.org/3/howto/logging.html#optimization) to use formatting in logging calls, as it creates an overhead to do the formatting, while using placeholders will make sure the formatting is delayed, and will only be applied if the log level means that the log method will be executed.

https://github.com/globality-corp/flake8-logging-format could be helpful to add in our flake8 plugins to enforce this.